
Quiclick is a python package provides analysis tools for RNA-seq and RNA editing analysis. We recommand exectue the files in an anaconda environment with python 3.x.

**Due to some reasons from author, this project will no longer upgrade.**

## 1 Download and install

Enter the following command in your terminal, then you can use command "quiclick" directly(NOT recommand). Because of the network of sourceforge, a Tsinghua cloud link is also provided: `https://cloud.tsinghua.edu.cn/f/c77843c493e64a2a9602/`, one can download a compressed package.

```shell
wget https://sourceforge.net/projects/quiclickrnaseq/files/quiclick.tar.gz
tar zxvf quiclick.tar.gz
```

Then use binary command to test whether the package is installed:

```shell
[username@localhost filename]python <quiclickPath>/quiclick/bin/quiclick.py
----------------------------------------------------------------------------------------------------------------
                                          Users help documentation
----------------------------------------------------------------------------------------------------------------

 Parameters         Description                                 Default & Note                    Optional(Y/N)
    -w           Pathway of raw data                             Current path                           Y
    -i           Pathway of index files                 Create automatically (ENSEMBL version)          Y
    -m        Multiprocessing for mapping                Single thread (supported files < 6)            Y
    -p          Threads for mapping                       16 (CANNOT use with -m altogether)            Y
    -E          RNA editing analysis model               Carry out RNA editing analysis                 Y
    -D  Whether to execute differential expression analysis             NO                              Y
    -k      Whether to save genome index files                     Do not save                          Y
    -s                Species                               N/A (necessary parameter)                   N

++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```



## 2 Parameters and basic usage

-w: 

The **ABSOLUTE PATH** of raw data, if not, the path will be set as current data. And the raw data file must be listed as the structure showed below:

```shell
rawdata/
├── a1
│   ├── a1_1.fq.gz
│   ├── a1_2.fq.gz
├── a2
│   ├── a2_1.fq.gz
│   ├── a2_2.fq.gz
├── b1
│   ├── b1_1.fq.gz
│   ├── b2_2.fq.gz
├── b2
│   ├── b2_1.fq.gz
│   ├── b2_2.fq.gz
|___
```

-i:

The **ABSOLUTE PATH** of index file, and must be indexed by STAR, what's more, fasta and gtf files need to be saved in the same directory. 

-m:

If -m is set, mapping process will be exectued in a multi-processing way, not supported when sample numbers > 6.

-p:

If not must-processing for mapping , you can set a thread number for your mapping processing, 20-50 is recommended.

-E:

If -E is set, RNA editing process will be executed.

-D:

If -D is set, differential expression analysis will be execute, note that you need to point out experimental samples and control samples, and the number of each group **MUST** >=2.

-k:

if -k is set, index files will be save in your work path with name "refs", you can keep it for your next time's analysis.

-s:

-s is the only nessary parameter to point out which species you want to analysis, only support human, mouse, zebrafish and C. elegans.

### USEAGE example:

Useage example of quiclick:

```shell
python quiclick.py -w /home/Andrew/rawdata/ -i /home/Andrew/index/ -s celegans -m -E -D -k
```

What's more, we here describe the usage of pyRED. pyRED is a python file including two functions for analytics of differential RNA editing sites and the co-variaty of RNA edting with certain variable, whose outputs are p-value and slope of the linear regression respectively. The detailed description of function is written in the functions' description.

```python
from pyRED import differentialEditingSiteEstimation
exprEditList = [1,1]
exprTotalList = [19,20]
ctrlEditList = [9,10]
ctrlEditList = [11，11]
EDTI = [1,1,9,10]
TOTAl = [19,20,11,11]
pValueOfExprToCtrlSample = differentialEditingSiteEstimation(exprEditList, exprTotalList, ctrlEditList, ctrlEditList, EDIT, TOTAL)

from pyRED import editingCovarEstimation
editCounts = [1,1,10]
totalCounts = [11,12,11]
covarsList = [1,2,3]
pValueOfCovars, slpoeOfCovars = editingCovarEstimation(editCounts, totalCounts, co)
```



## 3 Output

Output file structure will contain several new directories:

**refs**: containing index file generated by STAR.

**trimmed_data**: containing RNA-seq data formatted as .fq.gz, which is trimmed by trim_galore.

**mapping_output**: containing BAM file generated by STAR.

**RES_res**: containg RNA editing results.

**.count files**: generated by htseq-count, indicating the reads counts from each sample.

**Differential_Express_analysis.R**: used for differential expression analysis, the output will be saved in current path.

**NOTE**: File Differential_Express_analysis.R is used for differential expression analysis, exectue `Rscript Differential_Express_analysis.R` to get the ouput table and plots. And package `DESeq2` and `gglot2` need to be installed in advance. Meanwhile, binary file of quiclick is no longer provided, `pandas`,`numpy`,`matplotlib` pacakages need to be installed in your environment.

```shell
rawdata/
├── a1
│   ├── a1_1.fq.gz
│   ├── a1_2.fq.gz
│   ├── mappingProcess.log
│   ├── RES_res
│   │   ├── Gene_edit_count.txt
│   │   ├── PARAMETER.txt
│   │   ├── SPRINT_gene_annotation.res
│   │   ├── SPRINT_identified_all.res
│   │   ├── SPRINT_identified_hyper.res
│   │   ├── SPRINT_identified_regular.res
|   |   └── tmp
│   └── trimmed_data
│       ├── a1_1.fq.gz_trimming_report.txt
│       ├── a1_1_val_1.fq
│       ├── a1_1_val_1.fq.gz
│       ├── a1_2.fq.gz_trimming_report.txt
│       ├── a1_2_val_2.fq
│       ├── a1_2_val_2.fq.gz
│       └── editingProcess.log
├── DESeq2_res_order.txt
├── Differential_Express_analysis.R
├── Expressed_genes.txt
├── refs
└── Rplots.pdf
```

#### What's new
2020/9/27

Binary file of quiclick is no longer provided.

Added script: pyRED for RNA editing differential analysis. 



Contact: liuyf19@mails.tsinghua.edu.cn when questions arise.

